frontend/javascripts/admin/dataset/dataset_upload_view.js

```diff
@@ -63,30 +69,65 @@ class DatasetUploadView extends React.PureComponent<PropsWithForm, State> {
       // The original file object is contained in the originFileObj property
       // This is most likely not intentional and may change in a future Antd version
       formValues.zipFile = formValues.zipFile.map(wrapperFile => wrapperFile.originFileObj);
-      formValues.initialTeams = formValues.initialTeams.map(team => team.id);
 
       if (!err && activeUser != null) {
         Toast.info("Uploading datasets");
         this.setState({
           isUploading: true,
         });
 
-        const datasetConfig: DatasetConfig = {
-          ...formValues,
-          organization: activeUser.organization,
+        const datasetId: APIDatasetId = {
+          name: formValues.name,
+          owningOrganization: activeUser.organization,
         };
 
-        addDataset(datasetConfig).then(
-          async () => {
-            Toast.success(messages["dataset.upload_success"]);
-            trackAction("Upload dataset");
-            await Utils.sleep(3000); // wait for 3 seconds so the server can catch up / do its thing
-            this.props.onUploaded(activeUser.organization, formValues.name);
-          },
-          () => {
-            this.setState({ isUploading: false });
-          },
-        );
+        const resumableUpload = await createResumableUpload(datasetId, formValues.datastore);
+
+        resumableUpload.on("fileSuccess", file => {
+          this.setState({ isFinished: true });
+          const uploadInfo = {
+            uploadId: file.uniqueIdentifier,
+            organization: datasetId.owningOrganization,
+            name: datasetId.name,
+            initialTeams: formValues.initialTeams.map(team => team.id),
+          };
+
+          finishDatasetUpload(formValues.datastore, uploadInfo).then(
+            async () => {
+              Toast.success(messages["dataset.upload_success"]);
+              trackAction("Upload dataset");
+              await Utils.sleep(3000); // wait for 3 seconds so the server can catch up / do its thing
+              this.props.onUploaded(activeUser.organization, formValues.name);
+            },
+            () => {
+              this.setState({
+                isUploading: false,
+                isFinished: false,
+                isRetrying: false,
+                uploadProgress: 0,
+              });
+            },
+          );
+        });
+
+        resumableUpload.on("fileAdded", () => {
+          resumableUpload.upload();
+        });
```

Comment:
Instead of attaching this event handler, couldn't you move `resumableUpload.upload();` below the `addFiles` call in line 130? I'd assume that this is called only once (since only one file is added?). So, doing it explicitly without any listeners, seems more straightforward to me.

Maybe even `resumableUpload.addFiles(formValues.zipFile).upload();` would work as a shorthand.

Reply:
Sadly, this didn't work because `addFiles` is asynchrous and only finished when `fileAdded` is fired

Reply:
Ah, I see, then nevermind :)